RecoLocalTracker/Phase2TrackerRecHits/src/VectorHitBuilderAlgorithm.cc

```diff
@@ -0,0 +1,559 @@
+
+  //loop over the DetSetVector
+  LogDebug("VectorHitBuilderAlgorithm") << "with #clusters : " << ClustersPhase2Collection->size() << std::endl ;
+  for( auto DSViter : *ClustersPhase2Collection){
+
+    unsigned int rawDetId1(DSViter.detId());
+    DetId detId1(rawDetId1);
+    DetId lowerDetId, upperDetId;
+    if( theTkTopo->isLower(detId1) ){
+      lowerDetId = detId1;
+      upperDetId = theTkTopo->partnerDetId(detId1);
+    } else if (theTkTopo->isUpper(detId1)) {
+      upperDetId = detId1;
+      lowerDetId = theTkTopo->partnerDetId(detId1);
+    }
+    DetId detIdStack = theTkTopo->stack(detId1);
+
+    //debug
+    LogTrace("VectorHitBuilderAlgorithm") << "  DetId stack : " << detIdStack.rawId() << std::endl;
+    LogTrace("VectorHitBuilderAlgorithm") << "  DetId lower set of clusters  : " << lowerDetId.rawId();
+    LogTrace("VectorHitBuilderAlgorithm") << "  DetId upper set of clusters  : " << upperDetId.rawId() << std::endl;
+
+    it_temporary = tempVHAcc.find(detIdStack);
+    if ( it_temporary != tempVHAcc.end() ) {
+      LogTrace("VectorHitBuilderAlgorithm") << " this stack has already been analyzed -> skip it ";
+      continue;
+    }
+
+    const GeomDet* gd;
+    const StackGeomDet* stackDet;
+    edmNew::DetSetVector<Phase2TrackerCluster1D>::const_iterator it_detLower = ClustersPhase2Collection->find( lowerDetId );
+    edmNew::DetSetVector<Phase2TrackerCluster1D>::const_iterator it_detUpper = ClustersPhase2Collection->find( upperDetId );
+
+    if ( it_detLower != ClustersPhase2Collection->end() && it_detUpper != ClustersPhase2Collection->end() ){
+
+      gd = theTkGeom->idToDet(detIdStack);
+      stackDet = dynamic_cast<const StackGeomDet*>(gd);
+      std::vector<VectorHit> vhsInStack_Acc;
+      std::vector<VectorHit> vhsInStack_Rej; 
+      const auto vhsInStack_AccRej = buildVectorHits(stackDet, clusters, *it_detLower, *it_detUpper);
+
+      //storing accepted and rejected VHs
+      for(auto vh : vhsInStack_AccRej ) {
+        if(vh.second == true){
+          vhsInStack_Acc.push_back(vh.first);
+        }
+        else if(vh.second == false){
+          vhsInStack_Rej.push_back(vh.first);
+        }
+      }
+
+      //ERICA:: to be checked with map!
+      //sorting vhs for best chi2
+      std::sort(vhsInStack_Acc.begin(), vhsInStack_Acc.end());
+
+      tempVHAcc[detIdStack] = vhsInStack_Acc;
+      tempVHRej[detIdStack] = vhsInStack_Rej;
+
+      LogTrace("VectorHitBuilderAlgorithm") << "For detId #" << detIdStack.rawId() << " the following VHits have been accepted:";
+      for (auto vhIt : vhsInStack_Acc){
+        LogTrace("VectorHitBuilderAlgorithm") << "accepted VH: " << vhIt;
+      }
+      LogTrace("VectorHitBuilderAlgorithm") << "For detId #" << detIdStack.rawId() << " the following VHits have been rejected:";
+      for (auto vhIt : vhsInStack_Rej){
+        LogTrace("VectorHitBuilderAlgorithm") << "rejected VH: " << vhIt;
+      }
```

Comment:
Put the lines 227-234 inside `#ifdef EDM_ML_DEBUG` to avoid empty loops (even if compiler may optimize them away).